Sources/AsyncHTTPClient/ConnectionPool/HTTPRequestStateMachine.swift

```diff
@@ -0,0 +1,625 @@
+        case .failed:
+            return .wait
+        }
+    }
+
+    mutating func receivedHTTPResponseBodyPart(_ body: ByteBuffer) -> Action {
+        switch self.state {
+        case .initialized, .verifyingRequest, .waitForChannelToBecomeWritable:
+            preconditionFailure("How can we receive a response head before sending a request head ourselves. Invalid state: \(self.state)")
+
+        case .running(_, .waitingForHead):
+            preconditionFailure("How can we receive a response body, if we haven't received a head. Invalid state: \(self.state)")
+
+        case .running(let requestState, .receivingBody(let head, .downstreamHasDemand)):
+            self.state = .running(requestState, .receivingBody(head, .downstreamIsConsuming(readPending: false)))
+            return .forwardResponseBodyPart(body, resetReadTimeoutTimer: self.idleReadTimeout)
+
+        case .running(_, .receivingBody(_, .downstreamIsConsuming)):
+            // the state doesn't need to be changed. we are already in the correct state.
+            // just forward the data.
+            return .forwardResponseBodyPart(body, resetReadTimeoutTimer: self.idleReadTimeout)
+
+        case .running(_, .endReceived), .finished:
+            preconditionFailure("How can we successfully finish the request, before having received a head. Invalid state: \(self.state)")
+        case .failed:
+            return .wait
+        }
+    }
+
+    mutating func receivedHTTPResponseEnd() -> Action {
+        switch self.state {
+        case .initialized, .verifyingRequest, .waitForChannelToBecomeWritable:
+            preconditionFailure("How can we receive a response head before sending a request head ourselves. Invalid state: \(self.state)")
+
+        case .running(_, .waitingForHead):
+            preconditionFailure("How can we receive a response end, if we haven't a received a head. Invalid state: \(self.state)")
+
+        case .running(.streaming(let expectedBodyLength, let sentBodyBytes, let producerState), .receivingBody(let head, _)) where head.status.code < 300:
+            self.state = .running(
+                .streaming(expectedBodyLength: expectedBodyLength, sentBodyBytes: sentBodyBytes, producer: producerState),
+                .endReceived
+            )
+            return .wait
+
+        case .running(.streaming(_, _, let producerState), .receivingBody(let head, _)):
+            assert(head.status.code >= 300)
+            assert(producerState == .paused, "Expected to have paused the request body stream, when the head was received. Invalid state: \(self.state)")
+            self.state = .finished
+            return .succeedRequest(.close, clearReadTimeoutTimer: self.idleReadTimeout != nil)
+
+        case .running(.endSent, .receivingBody(_, let streamState)):
+            let finalAction: Action.FinalStreamAction
+            switch streamState {
+            case .downstreamIsConsuming(readPending: true):
+                finalAction = .read
```

Comment:
I do not understand why the consumer has any control over what we do with the HTTP connection at this point.

Reply:
I'm not sure I get your point here?

Reply:
Why does our `finalAction` depend on what the consumer is doing with the data?

Reply:
The `finalAction` depends on whether backpressure was used. If we receive the response end and we have a read event buffered, I think we should make sure this read event is passed forward. This way we ensure that the channel is in a "normal" state (no buffered read event), once we start the next request. I see it as a cleanup job.